x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/InternalGeoLine.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.geo;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.geo.GeoEncodingUtils;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A single line string representing a sorted sequence of geo-points
+ */
+public class InternalGeoLine extends InternalAggregation {
+    private static final Logger logger = LogManager.getLogger(InternalGeoLine.class);
+
+    private List<Position> line;
+
+    InternalGeoLine(String name, List<Position> line, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        this.line = line;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalGeoLine(StreamInput in) throws IOException {
+        super(in);
+        this.line = in.readList(Position::new);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeList(line);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return GeoLineAggregationBuilder.NAME;
+    }
+
+    public List<Position> line() {
+        return line;
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        builder.field("type", "LineString");
+        builder.array("coordinates", line.toArray());
+        return builder;
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
```

Comment:
Potential optimization: since the shard results are all pre-sorted, we could divide-conquer pairwise merge the lists instead of dumping them all into a single list and sorting.  Or a priority queue implementation to track the lowest item from each list and sort them in a streaming fashion.

Might be overkill though, we're only talking about something like 10*10k points